BE board, column and cards

ahmed-belhadj opened the pull request:
### What this PR does (required):
- [x] A user has one board. A board has many columns. A column has many cards
- [x] Cards need only to have a name and description. No comments, deadline, attachments, etc
- [x] A new board already comes with the "in progress" and "completed" columns and is created for the user when they sign up
- [x] Implement all routes for creating and editing columns within a board, creating cards, moving cards with a column and between columns

### Screenshots / Videos (required):
- Schemas: https://www.loom.com/share/41b1a18330ee4e9b9074ab952d4939fe
- Routes: https://www.loom.com/share/a8540ed3ea404454a7bc8d6a3509d3f7

### Any information needed to test this feature (required):
- ### POST localhost:3001/auth/register
  A new board already comes with the "in progress" and "completed" columns and is created for the user when they sign up
  ### Request:
  ```
  {
      "username": "testRegister",
      "email":"testRegister@test.com",
      "password": "password"
  }
  ```
  ### Response
  ```
    {
      "success": {
          "user": {
              "id": "60bd5ccd73db83085d86a27c",
              "username": "testRegister",
              "email": "testRegister@test.com",
              "board": {
                  "columns": [
                      {
                          "cards": [],
                          "_id": "60bd5ccc73db83085d86a279",
                          "title": "In Progress",
                          "__v": 0
                      },
                      {
                          "cards": [],
                          "_id": "60bd5ccc73db83085d86a27a",
                          "title": "Completed",
                          "__v": 0
                      }
                  ],
                  "_id": "60bd5ccc73db83085d86a27b",
                  "__v": 0
              }
          }
      }
  }
  ```

- ### POST localhost:3001/auth/login
  Upon login, full board with all column and card objects included
  ### Request:
  ```
  {
    "email": "testRegister@test.com",
    "password": "password"
  }
  ```
  ### Response:
  ```
      {
      "success": {
          "user": {
              "id": "60bd5ccd73db83085d86a27c",
              "username": "testRegister",
              "email": "testRegister@test.com",
              "board": {
                  "columns": [
                      {
                          "cards": [
                              {
                                  "_id": "60bd5d1f73db83085d86a27e",
                                  "name": "Math homework",
                                  "description": "March 10 2020",
                                  "__v": 0
                              },
                              {
                                  "_id": "60bd5dbb73db83085d86a280",
                                  "name": "Math finals",
                                  "description": "May 12",
                                  "__v": 0
                              }
                          ],
                          "_id": "60bd5cef73db83085d86a27d",
                          "title": "Math Todos",
                          "__v": 0
                      },
                      {
                          "cards": [],
                          "_id": "60bd5ccc73db83085d86a279",
                          "title": "In Progress",
                          "__v": 0
                      },
                      {
                          "cards": [],
                          "_id": "60bd5ccc73db83085d86a27a",
                          "title": "Completed",
                          "__v": 0
                      }
                  ],
                  "_id": "60bd5ccc73db83085d86a27b",
                  "__v": 0
              }
          }
      }
  }
  ```

- ### POST localhost:3001/column/create
  Create a column by board id
  ### Request:
  ```
    {
      "boardId":"60bd5ccc73db83085d86a27b",
      "column": {
          "title": "Math",
          "cards": []
      }
  }
  ```
  ### Response:
  ```
   {
      "cards": [],
      "_id": "60bd5cef73db83085d86a27d",
      "title": "Math",
      "__v": 0
  }
  ```
  
- ### POST localhost:3001/column/updateTitle
  Update a column's title by column id
  ### Request:
  ```
   {
      "columnId":"60bd5cef73db83085d86a27d",
      "columnTitle": "Math Todos"
  }
  ```
  ### Response:
  ```
  {
      "cards": [],
      "_id": "60bd5cef73db83085d86a27d",
      "title": "Math Todos",
      "__v": 0
  }
  ```
  
- ### POST localhost:3001/card/create
  Create a card by column id
  ### Request:
  ```
  {
      "card": {
          "name": "Math finals",
          "description": "May 12"
      },
      "columnId": "60bd5cef73db83085d86a27d"
  }
  ```
  ### Response:
  ```
  {
      "_id": "60bd5dbb73db83085d86a280",
      "name": "Math finals",
      "description": "May 12",
      "__v": 0
  }
  ```

- ### POST localhost:3001/card/updateName
  Update a card's name by card id
  ### Request:
  ```
   {
      "cardName":"Math homework",
      "cardId": "60bd5d1f73db83085d86a27e"
  }
  ```
  ### Response:
  ```
  {
      "_id": "60bd5d1f73db83085d86a27e",
      "name": "Math homework",
      "description": "May 10",
      "__v": 0
  }
  ```

- ### POST localhost:3001/card/updateDescription
  Update a card's description by card id
  ### Request:
  ```
  {
      "cardDescription":"March 10 2020",
      "cardId": "60bd5d1f73db83085d86a27e"
  }
  ```
  ### Response:
  ```
  {
      "_id": "60bd5d1f73db83085d86a27e",
      "name": "Math homework",
      "description": "March 10 2020",
      "__v": 0
  }
  ```

- ### GET localhost:3001/board
  Get board by id (includes all column and card objects)
  ### Request:
  ```
    {
      "boardId": "60bd5ccc73db83085d86a27b"
  }
  ```
  ### Response:
  ```
  {
      "columns": [
          {
              "cards": [],
              "_id": "60bd5ccc73db83085d86a279",
              "title": "In Progress",
              "__v": 0
          },
          {
              "cards": [
                  {
                      "_id": "60bd5d1f73db83085d86a27e",
                      "name": "Math homework",
                      "description": "March 10 2020",
                      "__v": 0
                  },
                  {
                      "_id": "60bd5dbb73db83085d86a280",
                      "name": "Math finals",
                      "description": "May 12",
                      "__v": 0
                  }
              ],
              "_id": "60bd5cef73db83085d86a27d",
              "title": "Math Todos",
              "__v": 0
          },
          {
              "cards": [],
              "_id": "60bd5ccc73db83085d86a27a",
              "title": "Completed",
              "__v": 0
          }
      ],
      "_id": "60bd5ccc73db83085d86a27b",
      "__v": 0
  }
  ```

- ### POST localhost:3001/board/update
  Update board ordering by sending new board
  ### Request:
  ```
   {
      "newBoard": {
          "columns": [
              {
                  "cards": [],
                  "_id": "60bd5ccc73db83085d86a279",
                  "title": "In Progress",
                  "__v": 0
              },
              {
                  "cards": [
                      {
                          "_id": "60bd5dbb73db83085d86a280",
                          "name": "Math finals",
                          "description": "May 12",
                          "__v": 0
                      },
                      {
                          "_id": "60bd5d1f73db83085d86a27e",
                          "name": "Math homework",
                          "description": "March 10 2020",
                          "__v": 0
                      }
                  ],
                  "_id": "60bd5cef73db83085d86a27d",
                  "title": "Math Todos",
                  "__v": 0
              },
              {
                  "cards": [],
                  "_id": "60bd5ccc73db83085d86a27a",
                  "title": "Completed",
                  "__v": 0
              }
          ],
          "_id": "60bd5ccc73db83085d86a27b",
          "__v": 0
      }
  }
  ```
  ### Response:
  ```
  {
      "columns": [
          {
              "cards": [],
              "_id": "60bd5ccc73db83085d86a279",
              "title": "In Progress",
              "__v": 0
          },
          {
              "cards": [
                  {
                      "_id": "60bd5dbb73db83085d86a280",
                      "name": "Math finals",
                      "description": "May 12",
                      "__v": 0
                  },
                  {
                      "_id": "60bd5d1f73db83085d86a27e",
                      "name": "Math homework",
                      "description": "March 10 2020",
                      "__v": 0
                  }
              ],
              "_id": "60bd5cef73db83085d86a27d",
              "title": "Math Todos",
              "__v": 0
          },
          {
              "cards": [],
              "_id": "60bd5ccc73db83085d86a27a",
              "title": "Completed",
              "__v": 0
          }
      ],
      "_id": "60bd5ccc73db83085d86a27b",
      "__v": 0
  }
  ```

### Any issues with the current functionality (optional):
- N/A

ahmed-belhadj committed:
Update package.json

add minimal eslint and prettier config

ahmed-belhadj committed:
Update package-lock.json

ahmed-belhadj committed:
Create .eslintrc.json

ahmed-belhadj committed:
Create .prettierrc

ahmed-belhadj committed:
Update User.js

A user has one board.

ahmed-belhadj committed:
Create Board.js

A board has many columns.

ahmed-belhadj committed:
Create Column.js

A column has many cards

ahmed-belhadj committed:
Create Card.js

Cards need only to have a name and description. No comments, deadline, attachments, etc

ahmed-belhadj committed:
Update auth.js

 A new board already comes with the "in progress" and "completed" columns and is created for the user when they sign up

ahmed-belhadj committed:
Update auth.js

return full board when user logs in

ahmed-belhadj committed:
Create column.js

create a column within a board by board id
update a column title by column id

ahmed-belhadj committed:
Create card.js

create a card within a column by column id
rename a card by card id
rename a card description by card id

ahmed-belhadj committed:
Create board.js

get full board (includes card and column objects ordered) by board id
reorder board by board id

ahmed-belhadj committed:
routers

boardRouter
columnRouter
cardRouter

ahmed-belhadj committed:
Update app.js

include routers

bonnieli commented on server/controllers/card.js:
```diff
@@ -0,0 +1,52 @@
+  }
+});
+
+// @route POST /card/updateName
```
Instead of having multiple endpoints to update different things, I suggest having one update endpoint (let's say /card/update for example, but technically you can also just create a PUT /card endpoint for best practices) and update validated attributes of any card

ahmed-belhadj replied:
Good point, I'll apply `/card/update` in this branch for now to make it available to the team today and then I'll use PUT in the upcoming BE: detailed card feature #2  for best practices

bonnieli reviewed (approved):
lgtm - your code looks clean but i do have that one commend on consolidating all your endpoints into one :)

ahmed-belhadj committed:
POST /card/update

Instead of having multiple endpoints to update different things, one update endpoint `/card/update` to update validated attributes of any card
